### PR TITLE
[SPARK-6913][SQL] Fixed "java.sql.SQLException: No suitable driver found" 

### PR DESCRIPTION
**Pull request by SlavikBaranov**
Fixed `java.sql.SQLException: No suitable driver found` when loading DataFrame into Spark SQL if the driver is supplied with `--jars` argument.

The problem is in `java.sql.DriverManager` class that can't access drivers loaded by Spark ClassLoader.

Wrappers that forward requests are created for these drivers.

Also, it's not necessary any more to include JDBC drivers in `--driver-class-path` in local mode, specifying in `--jars` argument is sufficient.

### CODE REVIEW

**Comment by rxin** [comment]
Jenkins, test this please.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/jdbc.scala [hunk 1]
```diff
@@ -179,4 +184,63 @@ package object jdbc {
     }
 
   }
+
+  private [sql] case class DriverWrapper(wrapped: Driver) extends Driver {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/jdbc.scala]: this doesn't need to be a case class, does it? Looks like you can just slightly change the pattern matching down below.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/jdbc.scala [hunk 2]
```diff
@@ -179,4 +184,63 @@ package object jdbc {
...
+    val lock = new ReentrantLock()
+
+    def register(className: String): Unit = {
+      val cls = Utils.getContextOrSparkClassLoader.loadClass(className)
+      if (cls.getClassLoader == null) {
+        logTrace(s"$className has been loaded with bootstrap ClassLoader, wrapper is not required")
+      } else if (wrapperMap.get(className).isDefined) {
+        logTrace(s"Wrapper for $className already exists")
+      } else {
+        lock.lock()
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/jdbc.scala]: given this is unlikely to be a performance bottleneck, why not just use a synchronized block?

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/jdbc.scala [hunk 3]
```diff
@@ -179,4 +184,63 @@ package object jdbc {
...
+  /**
+   * java.sql.DriverManager is always loaded by bootstrap classloader,
+   * so it can't load JDBC drivers accessible by Spark ClassLoader.
+   *
+   * To solve the problem, drivers from user-supplied jars are wrapped
+   * into thin wrapper.
+   */
+  private [sql] object DriverRegistry extends Logging {
+
+    val wrapperMap: mutable.Map[String, DriverWrapper] = mutable.Map.empty
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/jdbc.scala]: private

**Comment by SlavikBaranov** [comment]
Thanks for comments, fixed.

**Comment by rxin** [comment]
Jenkins, test this please.

**Comment by marmbrus** [comment]
This is very clever.  Thanks!  I ran the docker based mysql and postgres tests manually.  Merging to master.

**Comment by tgravescs** [comment]
@marmbrus @SlavikBaranov  @rxin this break the build with jdk6

[error] /home/tgraves/tgravescs_spark/sql/core/src/main/scala/org/apache/spark/sql/jdbc/jdbc.scala:198: value getParentLogger is not a member of java.sql.Driver
[error]     override def getParentLogger: java.util.logging.Logger = wrapped.getParentLogger
[error]                                                                      ^

java.sql.Driver.getParentLogger doesn't exist in jdk6, only jdk7

**Comment by tgravescs** [comment]
https://issues.apache.org/jira/browse/SPARK-7312

**Comment by marmbrus** [comment]
@tgravescs, thanks for the heads up.  We are taking a look and will revert the patch if there isn't an easy fix.

/cc @yhuai

**Comment by yhuai** [comment]
#5847 tries to fix it.

**Comment by wangxj1127** [comment]
Hi,
What is the workaround for this issue if I am using Spark 1.3.1? I have modified the compute-classpath.sh for appeding the mysql-connector jar, but still not working in yarn-cluster or yan-client mode. Only local mode works for me.

**Comment by wangxj1127** [comment]
Later on I found by adding "spark.executor.extraClassPath" property with value pointing to the jar works. But as Spark document below, the best way is to edit the compute-classpath.sh. Don't know why it doesn't work for me. 
https://spark.apache.org/docs/latest/sql-programming-guide.html#jdbc-to-other-databases

**Comment by Shawon91sust** [comment]
How to work the patch ? I am newbie in spark and facing the same problem. Can anyone tell me how to work this through. I mean how to resolve the suitable JDBC driver not found problem ....?

**Comment by wanshicheng** [comment]
> 你好，我是一个新手，遇到这样的问题不知道何从下手，希望可以得到有效的帮助，我是在使用sparkSQL将处理过的数据保存至mysql中，打包上传运行时出现的“java.sql.SQLException: No suitable driver”，spark版本是2.3.2，希望可以得到一个有效的解决办法，谢谢

在需要用到JDBC Driver的算子里，先注册一下Driver：
DriverRegistry.register(className)
它会根据全类名调用Spark的类加载器，加载Driver。

**Pull request closed without merging**